File: metamod/src/core/banner.cpp
#include "core/banner.h"

#include <algorithm>

namespace ch {
namespace {

const char* const kMark[] = {
    " ####   #   #",
    "#       #   #",
    "#       #####",
    "#       #   #",
    " ####   #   #",
};

constexpr size_t kMarkRows = sizeof(kMark) / sizeof(kMark[0]);

}  // namespace

std::vector<std::string> BuildBanner(const std::string& version,
                                     const std::string& platform, int serverId) {
    std::vector<std::string> info = {
        "ConnectHistory " + version,
        "Connection history and player analytics",
        std::string(),
        "Platform : " + platform,
        "Server   : #" + std::to_string(serverId),
    };

    size_t width = 0;
    for (const std::string& line : info) width = std::max(width, line.size());

    info[2] = std::string(width, '-');

    const size_t markWidth = std::string(kMark[0]).size();
    const std::string border = "+" + std::string(2 + markWidth + 3 + width + 2, '-') + "+";

    std::vector<std::string> lines;
    lines.push_back(border);
    for (size_t i = 0; i < kMarkRows; ++i) {
        std::string padded = info[i];
        padded.resize(width, ' ');
        lines.push_back("|  " + std::string(kMark[i]) + "   " + padded + "  |");
    }
    lines.push_back(border);

    return lines;
}

}  // namespace ch

File: cssharp/Utils/Banner.cs
using System;
using System.Collections.Generic;
using System.Globalization;

namespace ConnectHistory;

/// Заставка при загрузке плагина.
///
/// Рамка считается, а не прибита пробелами: версия приходит из тега и может быть
/// длиннее ожидаемого ("3.0.0-rc1"), а разъехавшаяся рамка в консоли сервера —
/// первое, что видит о плагине его владелец.
public static class Banner
{
    private static readonly string[] Mark =
    [
        " ####   #   #",
        "#       #   #",
        "#       #####",
        "#       #   #",
        " ####   #   #",
    ];

    public static IReadOnlyList<string> Build(string version, string platform, int serverId)
    {
        var info = new[]
        {
            "ConnectHistory " + version,
            "Connection history and player analytics",
            "",
            "Platform : " + platform,
            "Server   : #" + serverId.ToString(CultureInfo.InvariantCulture),
        };

        var width = 0;
        foreach (var line in info) width = Math.Max(width, line.Length);

        info[2] = new string('-', width);

        var inner = 2 + Mark[0].Length + 3 + width + 2;
        var border = "+" + new string('-', inner) + "+";

        var lines = new List<string>(Mark.Length + 2) { border };
        for (var i = 0; i < Mark.Length; i++)
            lines.Add("|  " + Mark[i] + "   " + info[i].PadRight(width) + "  |");
        lines.Add(border);

        return lines;
    }
}
